presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,424 @@
+                    .collect(toImmutableSet());
+            ImmutableList.Builder<Expression> joinPredicatesBuilder = ImmutableList.builder();
+
+            // add join conjuncts that were not used for inference
+            stream(EqualityInference.nonInferrableConjuncts(allFilter))
+                    .map(conjunct -> allInference.rewriteExpression(conjunct, symbol -> leftSymbols.contains(symbol) || rightSymbols.contains(symbol)))
+                    .filter(Objects::nonNull)
+                    // filter expressions that contain only left or right symbols
+                    .filter(conjuct -> allInference.rewriteExpression(conjuct, leftSymbols::contains) == null)
+                    .filter(conjuct -> allInference.rewriteExpression(conjuct, rightSymbols::contains) == null)
+                    .forEach(joinPredicatesBuilder::add);
+
+            // create equality inference on available symbols
+            // TODO: make generateEqualitiesPartitionedBy take left and right scope
+            List<Expression> joinEqualities = allInference.generateEqualitiesPartitionedBy(symbol -> leftSymbols.contains(symbol) || rightSymbols.contains(symbol)).getScopeEqualities();
+            EqualityInference joinInference = createEqualityInference(joinEqualities.toArray(new Expression[joinEqualities.size()]));
+            joinPredicatesBuilder.addAll(joinInference.generateEqualitiesPartitionedBy(in(leftSymbols)).getScopeStraddlingEqualities());
+
+            List<Expression> joinPredicates = joinPredicatesBuilder.build();
+            List<JoinNode.EquiJoinClause> joinConditions = joinPredicates.stream()
+                    .filter(JoinEnumerator::isJoinEqualityCondition)
+                    .map(predicate -> toEquiJoinClause((ComparisonExpression) predicate, leftSymbols))
+                    .collect(toImmutableList());
+            if (joinConditions.isEmpty()) {
+                return INFINITE_COST_RESULT;
+            }
+            List<Expression> joinFilters = joinPredicates.stream()
+                    .filter(predicate -> !isJoinEqualityCondition(predicate))
+                    .collect(toImmutableList());
+
+            Set<Symbol> requiredJoinSymbols = ImmutableSet.<Symbol>builder()
+                    .addAll(outputSymbols)
+                    .addAll(SymbolsExtractor.extractUnique(joinPredicates))
+                    .build();
+
+            JoinEnumerationResult leftResult = getJoinSource(
+                    ImmutableList.copyOf(leftSources),
+                    requiredJoinSymbols.stream().filter(leftSymbols::contains).collect(toImmutableList()));
+            if (leftResult.cost.hasUnknownComponents()) {
+                return UNKNOWN_COST_RESULT;
+            }
+            if (leftResult.cost.equals(INFINITE_COST)) {
+                return INFINITE_COST_RESULT;
+            }
+            PlanNode left = leftResult.planNode.orElseThrow(() -> new IllegalStateException("no planNode present"));
```

Comment:
Both `IllegalStateException` and `IllegalArgumentException` are very similar. Both of them are used to be thrown when the convention of calling a method is not followed. `IllegalArgumentException` is thrown when the inappropriate argument is passed, and `IllegalStateException` is thrown when a method is called on an object in an inappropriate state. 

What is actually done here - a simple assertion. An assertion that is supposed to check that the assumption about the cost and a plan node is correct. For such assertions it is better to use `verify`

`verify(leftResult.planNode.isPresent(), "planNode is not present")`
`PlanNode left = leftResult.planNode.get()`

Reply:
nit: or fluently: `PlanNode left = leftResult.planNode.orElseThrow(() -> new VerifyException("planNode is not present"));`